### PR TITLE
Feature/async rules reimplementation

### PR DESCRIPTION
**Pull request by coutoPL**
New experimental core. Typesafe, well-tested, backward compatible, immutable, implemented in Scala. Without LDAP rules (at the moment). Linked with ES 63x-experimental module. 

Existing ROR configs for old core should work as before, but if there is problem with config reloading it may mean that ROR found some bad configuration early - not on request handling level.

### CODE REVIEW

**Comment by sscarduzio** [comment]
Thank you @coutoPL! 
Can you expand on "but if there is problem with config reloading it may mean that ROR found some bad configuration early" and how this behaviour differs from previous?

**Comment by coutoPL** [comment]
eg. in old `JwtAuthAsyncRule`, in `match` method we find code like this:

`throw new RuntimeException("unrecognised algorithm family " + algoFamily + ". Should be either of: HMAC, EC, RSA, NONE");`

In new core this will be catched at configuration parsing level (not at request handling level). In old code, some misconfigurations was being handled too late. New core catches all such things as fast as they can be recognized. 

So, according to example above - user could set unknown algorithm and reload the config.
* old code behaviour:
 Config was parsed and new ACL engine was set. The user won't be informed about the problem until he sends first request.

* new core behaviour:
User will be informed about misconfiguration on config reload. New ACL engine won't be created.

File: es63x-experimental/src/main/java/tech/beshu/ror/es/ReadonlyRestPlugin.java [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+    implements ScriptPlugin, ActionPlugin, IngestPlugin, NetworkPlugin {
+
+  private final Settings settings;
+
+  private IndexLevelActionFilter ilaf;
+  private SettingsObservableImpl settingsObservable;
+  private Environment environment;
+
+  @Inject
+  public ReadonlyRestPlugin(Settings s, Path p) {
```
[the remark this answers is not in the document]
Comment (coutoPL) [reply es63x-experimental/src/main/java/tech/beshu/ror/es/ReadonlyRestPlugin.java]: won't fix

File: corex/src/test/scala/tech/beshu/ror/unit/acl/factory/decoders/GroupsRuleSettingsTests.scala [hunk 2]
```diff
@@ -0,0 +1,379 @@
...
+              |
+              |  users:
+              |  - username: cartman
+              |    groups: ["group1", "group3"]
+              |    auth_key: "cartman:pass"
+              |
+              |""".stripMargin,
+          assertion = errors => {
+            errors should have size 1
+            errors.head should be(RulesLevelCreationError(MalformedValue(
```
[the remark this answers is not in the document]
Comment (coutoPL) [reply corex/src/test/scala/tech/beshu/ror/unit/acl/factory/decoders/GroupsRuleSettingsTests.scala]: won't fix

File: corex/src/test/scala/tech/beshu/ror/unit/acl/factory/decoders/ExternalAuthorizationRuleSettingsTests.scala [hunk 3]
```diff
@@ -0,0 +1,831 @@
...
+              |      user_groups_provider: "GroupsService2"
+              |      groups: ["group3"]
+              |      users: user1
+              |
+              |  user_groups_providers:
+              |""".stripMargin,
+          httpClientsFactory = mockedHttpClientsFactory,
+          assertion = errors => {
+            errors should have size 1
+            errors.head should be(DefinitionsLevelCreationError(Message("user_groups_providers declared, but no definition found")))
```
[the remark this answers is not in the document]
Comment (coutoPL) [reply corex/src/test/scala/tech/beshu/ror/unit/acl/factory/decoders/ExternalAuthorizationRuleSettingsTests.scala]: won't fix

File: corex/src/test/scala/tech/beshu/ror/unit/acl/factory/decoders/ExternalAuthenticationRuleSettingsTests.scala [hunk 4]
```diff
@@ -0,0 +1,474 @@
...
+              |
+              |  - name: "ext1"
+              |    authentication_endpoint: "http://localhost:8080/auth1"
+              |    success_status_code: success
+              |    cache_ttl_in_sec: 60
+              |""".stripMargin,
+          httpClientsFactory = mockedHttpClientsFactory,
+          assertion = errors => {
+            errors should have size 1
+            errors.head should be(DefinitionsLevelCreationError(MalformedValue(
```
[the remark this answers is not in the document]
Comment (coutoPL) [reply corex/src/test/scala/tech/beshu/ror/unit/acl/factory/decoders/ExternalAuthenticationRuleSettingsTests.scala]: won't fix

**Comment by sscarduzio** [comment]
@coutopl can you fix the ci?

**Comment by coutoPL** [comment]
yes, I'll fix it today. OK, I'll try to add the experimental module to travis conf, but I have to exclude LDAP test for it.

**Comment by coutoPL** [comment]
Everything is done.
Nothing to add on my part.

**Comment by sscarduzio** [comment]
Thank you!

**Pull request merged**